modules/beefy/src/lib.rs

```diff
@@ -0,0 +1,855 @@
+		fn on_initialize(_n: T::BlockNumber) -> frame_support::weights::Weight {
+			<RequestCount<T, I>>::mutate(|count| *count = count.saturating_sub(1));
+
+			(0_u64)
+				.saturating_add(T::DbWeight::get().reads(1))
+				.saturating_add(T::DbWeight::get().writes(1))
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I>
+	where
+		BridgedBeefyMmrHasher<T, I>: 'static + Send + Sync,
+	{
+		/// Initialize pallet with BEEFY validator set and best finalized block number.
+		#[pallet::weight((T::DbWeight::get().reads_writes(2, 4), DispatchClass::Operational))]
+		pub fn initialize(
+			origin: OriginFor<T>,
+			init_data: InitializationDataOf<T, I>,
+		) -> DispatchResult {
+			ensure_owner_or_root::<T, I>(origin)?;
+
+			let init_allowed = !<BestBlockNumber<T, I>>::exists();
+			ensure!(init_allowed, <Error<T, I>>::AlreadyInitialized);
+
+			log::info!(target: "runtime::bridge-beefy", "Initializing bridge BEEFY pallet: {:?}", init_data);
+			Ok(initialize::<T, I>(init_data)?.into())
+		}
+
+		/// Halt or resume all pallet operations.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operational(origin: OriginFor<T>, operational: bool) -> DispatchResult {
+			ensure_owner_or_root::<T, I>(origin)?;
+			<IsHalted<T, I>>::put(!operational);
+
+			if operational {
+				log::info!(target: "runtime::bridge-beefy", "Resuming pallet operations");
+			} else {
+				log::warn!(target: "runtime::bridge-beefy", "Stopping pallet operations");
+			}
+
+			Ok(().into())
+		}
+
+		/// Change `PalletOwner`.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_owner(
+			origin: OriginFor<T>,
+			new_owner: Option<T::AccountId>,
+		) -> DispatchResultWithPostInfo {
+			ensure_owner_or_root::<T, I>(origin)?;
+			match new_owner {
+				Some(new_owner) => {
+					PalletOwner::<T, I>::put(&new_owner);
+					log::info!(target: "runtime::bridge-beefy", "Setting pallet Owner to: {:?}", new_owner);
+				},
+				None => {
+					PalletOwner::<T, I>::kill();
+					log::info!(target: "runtime::bridge-beefy", "Removed Owner of pallet.");
+				},
+			}
+
+			Ok(().into())
+		}
+
+		/// Submit commitment, generated by BEEFY validators.
+		///
+		/// Apart from the generic payload, the commitment contains the finalized (by BEEFY) block
+		/// number, so they must be always be imported in order. Importing commitment gives us
+		/// knowledge of header hash that has been finalized by BEEFY validators.
+		#[pallet::weight(0)] // TODO: compute weights
+		pub fn submit_commitment(
+			origin: OriginFor<T>,
+			// TODO: implement `TypeInfo` for `BridgedBeefySignedCommitment<T, I>`, `BeefyMmrProof`
+			// and `BridgedBeefyMmrLeafUnpacked::<T, I>`
+			encoded_commitment: Vec<u8>,
```

Comment:
While working on relay, I've found having mandatory `encoded_mmr_proof` and `mmr_leaf` arguments here breaks our basic assumptions: relay is not putting any restrictions on the node it is connected to. But in this case (unless I'm totally wrong), we need to be connected to node that has offchain indexing enabled and to generate a proof we need to have unpruned state (because it is generated during runtime call, iiuc).

The first requirement seems adequate, because it is how BEEFY is designed. The second one is redundant imo. So probably we shall:
1) make these two arguments optional, or even better - extract to a separate call;
2) get rid, or make optional this "unpacked" validator set structure. We shall be using merkle root of validator keys instead and optionally - save proved validator keys to the storage (to avoid verifying same keys next time when the same set is signing something).

Whether this will be fixed in this, or in follow up PRs, depends on how things will be going (review vs relay development).